src/include/duckdb/common/single_thread_pointer.h

```diff
@@ -0,0 +1,169 @@
+#include <cstdlib>
+
+template <typename T>
+class single_thread_ptr {
+public:
+	T *ptr;              // contained pointer
+	uint32_t *ref_count; // reference counter
+
+public:
+	// Default constructor, constructs an empty single_thread_ptr.
+	constexpr single_thread_ptr() : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct empty single_thread_ptr.
+	constexpr single_thread_ptr(std::nullptr_t) : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct a single_thread_ptr that wraps raw pointer.
+	explicit single_thread_ptr(T *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
+		}
+	}
+
+	single_thread_ptr(uint32_t *r, T *p) {
+		ptr = p;
+		ref_count = r;
+	}
+
+	template <class U>
+	single_thread_ptr(uint32_t *r, U *p) {
+		ptr = p;
+		ref_count = r;
+	}
+
+	// Construct a single_thread_ptr that wraps raw pointer of convertible type.
+	template <typename U>
+	single_thread_ptr(U *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
+		}
+	}
+
+	// Copy  constructor.
+	single_thread_ptr(const single_thread_ptr &sp) : ptr(nullptr), ref_count(nullptr) {
+		if (sp.ptr) {
+			ptr = sp.ptr;
+			ref_count = sp.ref_count;
+			++(*ref_count);
+		}
+	}
+
+	// Conversion constructor.
+	template <typename U>
+	single_thread_ptr(const single_thread_ptr<U> &sp) : ptr(nullptr), ref_count(nullptr) {
+		if (sp.ptr) {
+			ptr = sp.ptr;
+			ref_count = sp.ref_count;
+			++(*ref_count);
+		}
+	}
+
+	// move  constructor.
+	single_thread_ptr(single_thread_ptr &&sp) noexcept : ptr {sp.ptr}, ref_count {sp.ref_count} {
+		sp.ptr = nullptr;
+		sp.ref_count = nullptr;
+	}
+
+	// move  constructor.
+	template <class U>
+	single_thread_ptr(single_thread_ptr<U> &&sp) noexcept : ptr {sp.ptr}, ref_count {sp.ref_count} {
+		sp.ptr = nullptr;
+		sp.ref_count = nullptr;
+	}
+
+	// No effect if single_thread_ptr is empty or use_count() > 1, otherwise release the resources.
+	~single_thread_ptr() {
+		release();
+	}
+
+	void release() {
+		if (ptr && ref_count) {
+			if (--(*ref_count) == 0) {
+				delete ptr;
+			}
+		}
+		ptr = nullptr;
+	}
+
+	// Copy assignment.
+	single_thread_ptr &operator=(single_thread_ptr sp) noexcept {
+		std::swap(this->ptr, sp.ptr);
```

Comment:
Shouldn't copy assignment increase the reference count as well, or am I missing something?

Reply:
Here, the object will be created by the copy constructor as the argument is not a reference. the increase of reference count happens there.

Reply:
Ah got it, so essentially first the copy constructor is called, and then the copy assignment operator is called. Thanks!

Reply:
Any reason to use `std::swap` here instead of just assigning the values then?

Reply:
We need to make sure that the current object is deleted (the left part of the assignment). So by swapping, the sp turns into the left object and will be destructed automatically. Also, the std::swap saves one assignment per swap.

Reply:
That makes sense. Thanks!